{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pysam in /opt/conda/lib/python3.8/site-packages (0.16.0.1)\n",
      "\n",
      "First read:\n",
      " C0HVYACXX120402:7:1207:5722:57044\t1187\t20\t9483248\t27\t76M\t20\t9483381\t76\tTTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG\tarray('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])\t[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]\n",
      "\n",
      "First read parameters values: \n",
      "\n",
      "FLAG:  1187\n",
      "RNAME:  20\n",
      "POS:  9483248\n",
      "Mapping quality:  27\n",
      "CIGAR:  76M\n",
      "MRNM:  20\n",
      "MPOS:  9483381\n",
      "ISIZE:  209\n",
      "SEQ:  TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG\n",
      "QUAL:  array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])\n",
      "FLAG(read num.400000):  83\n",
      "FLAG(read num.800000):  83\n",
      "FLAG(read num.1200000):  163\n",
      "FLAG(read num.1600000):  163\n",
      "FLAG(read num.2000000):  1187\n",
      "FLAG(read num.2400000):  121\n",
      "FLAG(read num.2800000):  99\n",
      "\n",
      "Total number of unmapped reads:  17765\n",
      "Total number of reads:  2921629\n",
      "Total number of reads with mapping quality equal to 0:  126628\n",
      "Average mapping quality for all the reads: 55.91379158681681\n",
      "Average mapping quality for reads with mapping quality not equal to 0: 58.446975510921106\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!{sys.executable} -m pip install pysam\n",
    "import pysam\n",
    "\n",
    "aligned_file=pysam.AlignmentFile('merged-tumor.bam')\n",
    "num_unmapped=0\n",
    "num_zero_qual=0\n",
    "num_reads=0\n",
    "average_mapq=0\n",
    "average_mapq_nonzero=0\n",
    "\n",
    "for i,read in enumerate(aligned_file):\n",
    "    num_reads+=1\n",
    "    if i==0:\n",
    "        print(\"\\nFirst read:\\n\",read)\n",
    "        print(\"\\nFirst read parameters values: \\n\")\n",
    "        print(\"FLAG: \",read.flag)\n",
    "        print(\"RNAME: \",read.reference_id)\n",
    "        print(\"POS: \",read.reference_start)\n",
    "        print(\"Mapping quality: \",read.mapping_quality)\n",
    "        print(\"CIGAR: \",read.cigarstring)\n",
    "        print(\"MRNM: \",read.next_reference_id)\n",
    "        print(\"MPOS: \",read.next_reference_start)\n",
    "        print(\"ISIZE: \",read.template_length)\n",
    "        print(\"SEQ: \",read.query_sequence)\n",
    "        print(\"QUAL: \",read.query_qualities)\n",
    "            \n",
    "    if i%400000==0 and i>399999:\n",
    "        print(\"FLAG(read num.\"+str(i)+\"): \",read.flag)\n",
    "         \n",
    "    average_mapq+=read.mapping_quality\n",
    "    if read.is_unmapped:\n",
    "        num_unmapped+=1\n",
    "   \n",
    "    if read.mapping_quality==0:\n",
    "        num_zero_qual+=1\n",
    "    else:\n",
    "        average_mapq_nonzero+=read.mapping_quality\n",
    "    \n",
    "    \n",
    "average_mapq/=num_reads   \n",
    "average_mapq_nonzero/=(num_reads-num_zero_qual)\n",
    "\n",
    "\n",
    "print(\"\\nTotal number of unmapped reads: \",num_unmapped)\n",
    "print(\"Total number of reads: \",num_reads)\n",
    "print(\"Total number of reads with mapping quality equal to 0: \",num_zero_qual)\n",
    "print(\"Average mapping quality for all the reads:\",average_mapq)\n",
    "print(\"Average mapping quality for reads with mapping quality not equal to 0:\",average_mapq_nonzero)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First read parameters inspection:<br>\n",
    "1.FLAG=1187<br>\n",
    "  -read paired (0x1)<br>\n",
    "  -read mapped in proper pair (0x2)<br>\n",
    "  -mate reverse strand (0x20)<br>\n",
    "  -second in pair (0x80)<br>\n",
    "  -read is PCR or optical duplicate (0x400)<br>\n",
    "  \n",
    "2.RNAME=20 (Read name)\n",
    "\n",
    "3.POS=9483248 (Position of read alignment)\n",
    "\n",
    "4.Mapping quality=27\n",
    "\n",
    "5.CIGAR=76M (76 nucleotides are matched or mismatched)\n",
    "\n",
    "6.MRNM=20 (Name of mate)\n",
    "\n",
    "7.MPOS=9483381(Position of mate)\n",
    "\n",
    "\n",
    "8.ISIZE=209 (Template length)\n",
    "\n",
    "\n",
    "9.SEQ=TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG (Sequence of nucleotides in this read)\n",
    "\n",
    "10.QUAL=array('B',[28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])\n",
    "(Quality scores and its not ASCII-encoded value as it is used in BAM files so there is no offset 33 that needs to be substracted)\n",
    "\n",
    "FLAG inspection:\n",
    "\n",
    "Read number 400000 and 800000:<br>\n",
    "FLAG=83<br>\n",
    "-read paired (0x1)<br>\n",
    "-read mapped in proper pair (0x2)<br>\n",
    "-read reverse strand (0x10)<br>\n",
    "-first in pair (0x40)<br>\n",
    "\n",
    "Read number 1200000 and 1600000:<br>\n",
    "FLAG=163<br>\n",
    "-read paired (0x1)<br>\n",
    "-read mapped in proper pair (0x2)<br>\n",
    "-mate reverse strand (0x20)<br>\n",
    "-second in pair (0x80)<br>\n",
    "\n",
    "Read number 2400000:<br>\n",
    "FLAG=121<br>\n",
    "-read paired (0x1)<br>\n",
    "-mate unmapped (0x8)<br>\n",
    "-read reverse strand (0x10)<br>\n",
    "-mate reverse strand (0x20)<br>\n",
    "-first in pair (0x40)<br>\n",
    "\n",
    "Read number 2800000:<br>\n",
    "FLAG=99<br>\n",
    "-read paired (0x1)<br>\n",
    "-read mapped in proper pair (0x2)<br>\n",
    "-mate reverse strand (0x20)<br>\n",
    "-first in pair (0x40)<br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
